src/Features/CSharp/Portable/ConvertLinq/CSharpConvertForEachToLinqQueryProvider.cs

```diff
@@ -0,0 +1,585 @@
+
+        // Do not try to refactor queries with comments or conditional compilation in them.
+        // We can consider supporting queries with comments in the future.
+        protected override bool Validate(StatementSyntax statement)
+         => !(statement.ContainsDirectives ||
+            statement.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+             SyntaxKind.SingleLineCommentTrivia,
+             SyntaxKind.MultiLineCommentTrivia,
+             SyntaxKind.MultiLineDocumentationCommentTrivia)));
+
+        protected override (ImmutableArray<SyntaxNode> ConvertingNodes, ImmutableArray<SyntaxToken> Identifiers, ImmutableArray<StatementSyntax> Statements) CreateForEachInfo(ForEachStatementSyntax forEachStatement)
+        {
+            var identifiers = new List<SyntaxToken>();
+            identifiers.Add(forEachStatement.Identifier);
+            var convertingNodes = new List<SyntaxNode>();
+            var current = forEachStatement.Statement;
+            IEnumerable<StatementSyntax> statementsCannotBeConverted = null;
+
+            void ProcessLocalDeclarationStatement(LocalDeclarationStatementSyntax localDeclarationStatement)
+            {
+                foreach (var variable in localDeclarationStatement.Declaration.Variables)
+                {
+                    convertingNodes.Add(variable);
+                    identifiers.Add(variable.Identifier);
+                }
+            }
+
+            // Setting statementsCannotBeConverted to anything means that we stop processing.
+            while (statementsCannotBeConverted == null)
+            {
+                switch (current.Kind())
+                {
+                    case SyntaxKind.Block:
+                        var block = (BlockSyntax)current;
+                        var array = block.Statements.ToArray();
+                        if (array.Any())
+                        {
+                            // Process all statements except the last one.
+                            for (int i = 0; i < array.Length - 1; i++)
+                            {
+                                var statement = array[i];
+                                if (statement is LocalDeclarationStatementSyntax localDeclarationStatement)
+                                {
+                                    ProcessLocalDeclarationStatement(localDeclarationStatement);
+                                }
+                                else
+                                {
+                                    statementsCannotBeConverted = array.Skip(i);
+                                    break;
+                                }
+                            }
+
+                            // Process the last statement separately.
+                            current = array.Last();
+                        }
+                        else
+                        {
+                            statementsCannotBeConverted = Enumerable.Empty<StatementSyntax>();
```

Comment:
this is a bit subtle.  You're depending on setting this to any value (even an **empty** list) breaking the outer loop...

Reply:
yes, technically the method signature could be:
StatementSyntax[] ProcessWhileCanConvert(... out identifiers, out nodeForConvcersion, out ..., out .....)
However, StatementSyntax is not the only value we get from the method.

The method performs a traverse of nested statements inside a given foreach.
As soon as we find something that cannot be translated into a linq clause, we set the rest into statementsCannotBeConverted and stop processing.

---
In reply to: [191058220](https://github.com/dotnet/roslyn/pull/27130#discussion_r191058220) [](ancestors = 191058220)